libsolidity/analysis/TypeChecker.cpp

```diff
@@ -920,6 +913,10 @@ bool TypeChecker::visit(Conditional const& _conditional)
 	}
 
 	_conditional.annotation().type = commonType;
+	_conditional.annotation().isPure =
+		_conditional.condition().annotation().isPure &&
```

Comment:
Ah, now I see where you are going :-)
On the other hand, things like `if (false) { .... }` are kind of useful for e.g. disabling warnings.